azure-iot-device/tests/common/pipeline/pipeline_stage_test.py

```diff
@@ -14,29 +14,202 @@
     assert_callback_failed,
     get_arg_count,
     add_mock_method_waiter,
-    StageTestBase,
     StageRunOpTestBase,
     StageHandlePipelineEventTestBase,
 )
 from azure.iot.device.common.pipeline.pipeline_stages_base import PipelineStage, PipelineRootStage
-from tests.common.pipeline.pipeline_event_test import add_instantiation_test
 from azure.iot.device.common.pipeline import pipeline_thread, pipeline_exceptions
 from azure.iot.device.common import handle_exceptions
-from .fixtures import FakeOperation
-from .helpers import assert_callback_succeeded
+from .fixtures import ArbitraryOperation
 
 logging.basicConfig(level=logging.DEBUG)
 
 
 def add_base_pipeline_stage_tests(
```

Comment:
>add_base_pipeline_stage_tests [](start = 4, length = 29)

With this change, we should be able to remove the monkeypatching from the base pipeline stage tests. Should we?  (I'd be willing to try it as an experiment)

Reply:
As discussed offline, I still feel that the monekypatching is superior here for safety - there are between 4-6 different test units (i.e. classes) added depending on the class, and you'd have to remember to import all 4-6 parent classes, and then subclass them (usually with no additional tests), which could lead to easily forgotten tests. Furthermore, the monkeypatch adds dynamic logic that only sometimes adds certain test units.

---
In reply to: [354873351](https://github.com/Azure/azure-iot-sdk-python/pull/409#discussion_r354873351) [](ancestors = 354873351)